src/dpapimkv1_fmt_plug.c

```diff
@@ -0,0 +1,436 @@
+
+	hmacSalt = out;
+	expected_hmac = out + 16;
+	last_key = out + (cur_salt->encrypted_len / 2) - 64;
+
+	hmac_sha1(pwdhash, 32, hmacSalt, 16, encKey, DIGEST_LEN);
+	hmac_sha1(encKey, DIGEST_LEN, last_key, 64, computed_hmac, DIGEST_LEN);
+
+	return memcmp(expected_hmac, computed_hmac, DIGEST_LEN);
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index = 0;
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+	for (index = 0; index < count; index += MAX_KEYS_PER_CRYPT)
+	{
+
+#ifdef SIMD_COEF_32
+		int lens[MAX_KEYS_PER_CRYPT];
+		unsigned char *pin[MAX_KEYS_PER_CRYPT];
+		union {
+			uint32_t *pout[MAX_KEYS_PER_CRYPT];
+			unsigned char *poutc;
+		} x;
+#endif
+
+		unsigned char *passwordBuf;
+		int passwordBufSize, digestlen = 20;
+		unsigned char out[MAX_KEYS_PER_CRYPT][KEY_LEN + IV_LEN];
+		unsigned char out2[MAX_KEYS_PER_CRYPT][KEY_LEN + IV_LEN];
+
+		SHA_CTX ctx;
+		MD4_CTX ctx2;
+
+		int i;
+
+		for (i = 0; i < MAX_KEYS_PER_CRYPT; ++i) {
+			passwordBuf = (unsigned char*)saved_key[index+i];
+			passwordBufSize = strlen16((UTF16*)passwordBuf) * 2;
+			
+			/* local credentials */
+			if (cur_salt->cred_type == 1)
+			{
+				SHA1_Init(&ctx);
+				SHA1_Update(&ctx, passwordBuf, passwordBufSize);
+				SHA1_Final(out[i], &ctx);
+				digestlen = 20;
+			}
+			/* domain credentials */
+			else if (cur_salt->cred_type == 2)
+			{
+				MD4_Init(&ctx2);
+				MD4_Update(&ctx2, passwordBuf, passwordBufSize);
+				MD4_Final(out[i], &ctx2);
+				digestlen = 16;
+			}
+
+			passwordBuf = (unsigned char*)cur_salt->SID;
+			passwordBufSize = (strlen16(cur_salt->SID) + 1) * 2;
+			hmac_sha1(out[i], digestlen, passwordBuf, passwordBufSize, out2[i], 20);
+#ifdef SIMD_COEF_32
+			lens[i] = 20;
+			pin[i] = (unsigned char*)out2[i];
+			x.pout[i] = (uint32_t*)(out[i]);
+#endif
+			}
+#ifdef SIMD_COEF_32
+		pbkdf2_sha1_sse((const unsigned char **)pin, lens, cur_salt->iv, MAX_IV_LEN, cur_salt->pbkdf2_iterations, &(x.poutc), KEY_LEN + IV_LEN, 0);
+#else
+		pbkdf2_sha1(out2[0], 20, cur_salt->iv, 16, cur_salt->pbkdf2_iterations, out[0], KEY_LEN + IV_LEN, 0);
+#endif
+
+		/* decrypt will use 32 bytes, we only initialized 20 so far */
+		for (i = 0; i < MAX_KEYS_PER_CRYPT; ++i) {
+			memset(out2[i] + 20, 0, 32 - 20);
+
+			if (decrypt(out[i], out[i] + KEY_LEN, out2[i], cur_salt->encrypted) == 0)
+				cracked[index+i] = 1;
+			else
+				cracked[index+i] = 0;
+		}
+	}
+	return count;
+}
+
+static int cmp_all(void *binary, int count)
+{
+	int index;
+	for (index = 0; index < count; index++)
+		if (cracked[index])
+			return 1;
+	return 0;
+}
+
+static int cmp_one(void *binary, int index)
+{
+	return cracked[index];
+}
+
+static int cmp_exact(char *source, int index)
+{
+    return 1;
+}
+
+static void dpapimkv1_set_key(char *key, int index)
+{
+	/* Convert key to UTF-16LE (--encoding aware) */
+	enc_to_utf16(saved_key[index], PLAINTEXT_LENGTH, (UTF8*)key, strlen(key));
+}
```

Comment:
Oh, so the credit bounces back to me because I brute forced that into his head over several years 😆